### PR TITLE
Ensure we treat "type assertions" as JSX within a unary expression

### LINKED ISSUE #50675: Debug failure at walk
# Bug Report

Running `tsc` causes an error message:

```
|| Error: Debug Failure.
||     at walk (/Users/derek/Workspace/***/node_modules/typescript/lib/tsc.js:98559:38)
||     at Object.forEachChildRecursively (/Users/derek/Workspace/***/node_modules/typescript/lib/tsc.js:25491:27)
||     at /Users/derek/Workspace/***/node_modules/typescript/lib/tsc.js:98475:20
||     at runWithCancellationToken (/Users/derek/Workspace/***/node_modules/typescript/lib/tsc.js:98388:24)
||     at getJSSyntacticDiagnosticsForFile (/Users/derek/Workspace/***/node_modules/typescript/lib/tsc.js:98472:20)
||     at getSyntacticDiagnosticsForFile (/Users/derek/Workspace/***/node_modules/typescript/lib/tsc.js:98380:65)
||     at /Users/derek/Workspace/***/node_modules/typescript/lib/tsc.js:98339:24
||     at Object.flatMap (/Users/derek/Workspace/***/node_modules/typescript/lib/tsc.js:428:25)
||     at getDiagnosticsHelper (/Users/derek/Workspace/***/node_modules/typescript/lib/tsc.js:98335:56)
||     at Object.getSyntacticDiagnostics (/Users/derek/Workspace/***/node_modules/typescript/lib/tsc.js:98343:20)
```

### 🔎 Search Terms

Debug failure

### 🕗 Version & Regression Information

- This is a crash
- This changed between versions 4.3.5 and 4.8.2

### 💻 Code

>`tsconfig.js`
```
{
  "extends": "expo/tsconfig.base",
  "compilerOptions": {
    "sourceMap": true,
    "strict": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "baseUrl": ".",
    "paths": {
      "@/*": ["src/*"]
    }
  },
  "exclude": ["jest.config.js", "babel.config.js", "app.config.js"],
  "ts-node": {
    "require": ["tsconfig-paths/register"],
    "transpileOnly": true,
    "module": "commonjs",
    "target": "es2015",
    "esModuleInterop": true
  }
}
```

### 🙁 Actual behavior

Running `tsc` gives a `Debug failure` error.

### 🙂 Expected behavior

Running `tsc` should have worked without an error

### PR DESCRIPTION
**Pull request by jakebailey**
Fixes #50675

~The comment claims we won't parse these in JS files, but that's not true; we can pretty easily parse them just like we do `as` and should error on them just like we do `as`.~

See: https://github.com/microsoft/TypeScript/pull/52667#issuecomment-1423071408

### CODE REVIEW

**Comment by fatcerberus** [comment]
I seem to recall a change was made a year or so ago to deliberately not parse legacy type assertions in JS in order to resolve ambiguity with expressions like `x < y > z`

**Comment by fatcerberus** [comment]
Unfortunately, I can’t seem to find the issue/PR for that now.  I *did* find several comments— including one from Ryan—suggesting JSX is always parsed in `.js` files though, which would indeed be ambiguous with old-style type assertions.

**Comment by jakebailey** [comment]
My impression was that we conditionalized a lot of the parser on the file extension, including JSX stuff. But that's still a good point, and I need to go figure out if this production should even be parsed.

**Comment by jakebailey** [comment]
So yep, you're right; we do generally treat JS as JSX. Going though this, though, here's what's happening; given:

```ts
const foo = <number> bar
```

Here, we see that there's a `<`, we say "yes, this is an update expression", then we go down and decide that since we're JSX, we parse `<number>` as JSX.

But, if we instead stick a unary operator at the front, like:

```
const foo = +<number> bar
```

We see that it's not an update expression, so we parse a "simple" unary expression, but as the parser currently operates, a simple unary expression only ever parses more simple unary expressions afterwards, i.e. `void +foo` and such. But since `<number> bar` is a unary operator, we parse `+ <number> bar` and end up with the type assertion which isn't possible at the top level.

I'm pretty sure that we need to be parsing JSX here as you've said, so, thanks for the catch. That's what [esbuild does too](https://jakebailey.github.io/esbuild-playground/#Ly8gQGZpbGVuYW1lOiBjb25maWcuanNvbgp7CiAgImZvcm1hdCI6ICJlc20iCn0KCi8vIEBmaWxlbmFtZTogaW5kZXgudHN4CmNvbnN0IHggPSAxMjM0OwoKZXhwb3J0IGZ1bmN0aW9uIEFwcCgpIHsKICByZXR1cm4gKyA8bnVtYmVyPiB4Owp9Cg==) (try changing the extension around between ts/js/tsx/jsx, and I think that's correct.

**Comment by jakebailey** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test tsserver top100
@typescript-bot run dt
@typescript-bot perf test this
@typescript-bot pack this

**Review by DanielRosenwasser — APPROVED** [review]

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -6163,6 +6168,7 @@ namespace Parser {
     }
 
     function parseTypeAssertion(): TypeAssertion {
+        Debug.assert(scriptKind === ScriptKind.TS || scriptKind === ScriptKind.TSX, "Type assertions should never be parsed outside of TS/TSX; they should either be comparisons or JSX.");
```
Comment (jakebailey) [inline src/compiler/parser.ts:6171]: On second reading, I this should just be `languageVariant !== LanguageVariant.JSX`? We can't actually parse out a `TypeAssertion` in TSX either.

**Review by andrewbranch — APPROVED** [review]

**Pull request merged**